### ISSUE TITLE
[TRACKER] FBX import issues

### ISSUE THREAD

**Issue by RevoluPowered** [godotengine/godot#47269]
So it's time to measure the problem areas of the FBX importer to make it more robust with more files.

FBX Parser / where the structure really falls down for us right now
- (needs) memory management is bad in places really needs a singular cleanup step at the end, not with shared ptr as right now the behaviour is buggy with parsing errors
- (idea stage) We could use clang libfuzzer to detect and cause problems with the parser / crashes / leaks etc. thanks @hpvb 
- (needs) FBX properties applied directly to the class, not via a ptr (refactor)
- (needs) bug fix for Limb vs LimbNode node reported by @W4RH4WK
- (needs) code cleanup, and upgrade to better practices, ideally we break it more now and make it more robust.
- (needs) rewrite of the FBX Video texture handling, it has been a big cause of crashes for a long time in assimp and here and its important it is fixed properly because we keep having misc issues with different files with this, for edge cases it seems built wrong.

FBX File Support
- (required) Improve the skin algorithm further with @reduz's suggestion to do something like this
```cpp
// better alternative to supporting global skins in godot by reduz.
// Step 1
Transform mesh_localizer = skeleton.global_transform.affine_inverse() * mesh.global_transform
mesh.apply_transform(mesh_localizer); //this applies to all vertices, normals, tangents
mesh.set_new_parent(skeleton);
mesh.set_transform(Transform());
// you will need to multiply all vertices by the mesh_localizer transforms. For normals and tangents, you need to use // mesh_localizer.basis.affine_inverse().inv_xform(normal)
// remember that normal transform is basis.affine_inverse().transposed()
// (as in, to transform a normal you need to do multiplication by inverse transposed)
// this is a general rule when transforming normals
// in Godot (and coincidentially GLTF), skeleton transform is applied first
// because its what makes the most sense
// for this to actually happen, meshes must have no transform
// since you go meshvertex -> vertex transform -> world
// for skinning to work, meshes should have no offset in regard to the skeleton
// you are basically getting rid of the difference in transform between the mesh and the skeleton
```

- (required) 3DS max pivot support (to fix weird rotation animations) 
- (required) cinema4d implementation

Maya support:
- so far nothing scary I have seen should be implemented.

Godot support required:
- nothing to report 

Blender fixes required:
- some blender skeletons export non uniformly even after #46525 
- this breaks compat with maya which means I won't merge it until I set the skin to be one formula for blender and another for maya.
- I should discuss with blender how to resolve this, but it could be a bug with vertex weight indexing.
- this PR needs finished off.

Unit tests:
- We need to make test models in max like we did for maya to get good support
- Euler unit tests should be re-added when we made them for 4.0

Issues in the tracker to fix complemented by the above: 
- #46906 File version 7300, probably made in 3DS max / solved with pivot formula fix
- #45573 Cinema4D
- #47174 honestly sometimes the values here are suspicious so I'd like to know we are getting the right values here -> unit tests
- #47184 FBX Video Texture rewrite
- #43816 Memory leaks, blender FBX files will be fixed by above tracker
- #43825 will look into this again, to work out if it needs closed or kept open.
- #46525 This will be finished off in this work.
- https://github.com/godotengine/godot/issues/45930
- https://github.com/godotengine/godot/issues/53544
- https://github.com/godotengine/godot/issues/55592

**Comment by BastiaanOlij** [comment]
Sweet! Looks like a good overview @RevoluPowered , any ideas what would be say the top 5 priorities to deal with first? I've got a personal stake in the global skinning issue as you know so I'm biassed towards that one :)

**Comment by likeich** [comment]
Should #45930 be included here?

**Comment by RevoluPowered** [comment]
> Sweet! Looks like a good overview @RevoluPowered , any ideas what would be say the top 5 priorities to deal with first? I've got a personal stake in the global skinning issue as you know so I'm biassed towards that one :)

1. (required) Improve the skin algorithm further with @reduz's suggestion to do something like this
2. FBX Parser Fixes - refactoring, and any memory bugs, also if there are keyword's missing in the parser we should fix them.
3. Any outstanding bugs in animations with maya - which have regressed in the master branch significantly. (some rotation initialisations have been broken from 3.2.x to master)
4. 3DS Max pivot formula needs implemented and then we need to add to the skin fix, to apply the geometric pivot direct to the mesh to avoid the value being calculated into the nodes.
5. I'd do a stabilisation pass to fix crashes and bugs reported.

> Should #45930 be included here?

It really depends, since I've not looked at the problem directly to reproduce it, I will check if it fits in the constraints for this, as once we make these fixes it should make a lot of reliability problems go away, so it could end up fixed by the above anyway.

**Comment by AgentSmith0** [comment]
I also think https://github.com/godotengine/godot/issues/47269 should be added to this tracker. The problem is now reproducable.

**Comment by Calinou** [comment]
@AgentSmith0 Done.

**Comment by akien-mga** [comment]
This is probably fixed with the new FBX importer using [FBX2glTF](https://github.com/V-Sekai/FBX2glTF/releases).
You can test it in Godot 4.0 alpha 6: https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-6
This new importer might also be backported for the `3.x` branch eventually.

If you still find issues with these models using the new importer, we suggest opening a new issue with details about the problems in Godot 4.0 alpha 6 or later.
